libs/vscode/messaging/src/lib/messaging-server.ts

```diff
@@ -0,0 +1,95 @@
+import net from 'net';
+import { ExtensionContext } from 'vscode';
+import { getFullOsSocketPath, killSocketOnPath } from './pipe';
+import { NxTerminalMessage } from './features/terminal-message';
+
+import { createMessageConnection } from 'vscode-jsonrpc/node';
+import {
+  NxEndedRunningTasks,
+  NxStartedRunningTasks,
+  NxUpdatedRunningTasks,
+} from './features/running-tasks';
+import {
+  MessagingNotification,
+  MessagingNotification2,
+} from './messaging-notification';
+
+const notifications: Array<MessagingNotification | MessagingNotification2> = [
+  NxTerminalMessage,
+  NxStartedRunningTasks,
+  NxEndedRunningTasks,
+  NxUpdatedRunningTasks,
+];
+
+export class NxMessagingServer {
+  #server: net.Server;
+  #fullSocketPath: string;
+
+  constructor(workspacePath: string, socketPath: string) {
+    this.#fullSocketPath = socketPath;
+    this.#server = net.createServer((socket) => {
+      console.log(`Client connected`);
+
+      // Create a connection for the server over this socket
+      const connection = createMessageConnection(
+        socket, // Use the TCP socket for communication
+        socket,
+      );
+
+      notifications.forEach((notification) => {
+        if ('type' in notification) {
+          connection.onNotification(
+            notification.type.method,
+            notification.handler,
+          );
+        }
+      });
+
+      connection.listen();
+
+      socket.on('close', () => {
+        connection.dispose();
+        console.log('Client disconnected');
+      });
+    });
+  }
+
+  listen() {
+    killSocketOnPath(this.#fullSocketPath);
+    this.#server.listen(this.#fullSocketPath, () => {
+      console.log(
+        `Nx Console Messaging JSON-RPC server listening on ${this.#fullSocketPath}`,
+      );
+    });
+  }
+
+  dispose() {
+    try {
+      this.#server.close(() => {
+        console.log(
+          `Nx Console Messaging JSON-RPC server closed on ${this.#fullSocketPath}`,
+        );
+      });
+    } catch (error) {
+      console.error('Error closing server:', error);
+    }
+  }
+}
+
+let existingServer: NxMessagingServer | null = null;
+
+export async function initMessagingServer(
+  context: ExtensionContext,
+  workspacePath: string,
+) {
+  if (existingServer) {
+    existingServer.dispose();
+  }
+  const socketPath = await getFullOsSocketPath(workspacePath);
+  const messagingServer = new NxMessagingServer(workspacePath, socketPath);
```

Comment:
what happens if the workspace root changes? It's probably overall fine if the socket keeps being in another folder but maybe we should handle that case...

Reply:
I guess the TUI connects via the workspace path so it might break in that case